File: src/spacetraders/domain/shared/contract.py
"""Contract domain entity and value objects"""
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import List

from .value_objects import Waypoint
from .exceptions import DomainException


class ContractException(DomainException):
    """Base exception for contract-related errors"""
    pass


class ContractAlreadyAcceptedError(ContractException):
    """Raised when trying to accept an already accepted contract"""
    pass


class ContractStatus(Enum):
    """Contract status enumeration"""
    OFFERED = "OFFERED"
    ACCEPTED = "ACCEPTED"
    FULFILLED = "FULFILLED"
    FAILED = "FAILED"
    EXPIRED = "EXPIRED"


@dataclass(frozen=True)
class Payment:
    """Payment terms for a contract"""
    on_accepted: int  # Credits received when accepting contract
    on_fulfilled: int  # Credits received when fulfilling contract

    def total(self) -> int:
        """Calculate total payment"""
        return self.on_accepted + self.on_fulfilled


@dataclass(frozen=True)
class Delivery:
    """Delivery requirement for a contract"""
    trade_symbol: str  # Good to deliver (e.g., "IRON_ORE")
    destination: Waypoint  # Where to deliver
    units_required: int  # Total units required
    units_fulfilled: int  # Units already delivered

    def __post_init__(self):
        """Validate delivery data"""
        if self.units_required < 0:
            raise ValueError("units_required cannot be negative")
        if self.units_fulfilled < 0:
            raise ValueError("units_fulfilled cannot be negative")
        if self.units_fulfilled > self.units_required:
            raise ValueError("units_fulfilled cannot exceed units_required")

    def remaining(self) -> int:
        """Calculate remaining units to deliver"""
        return self.units_required - self.units_fulfilled

    def is_fulfilled(self) -> bool:
        """Check if delivery requirement is met"""
        return self.units_fulfilled >= self.units_required


@dataclass(frozen=True)
class ContractTerms:
    """Terms and conditions of a contract"""
    deadline: datetime  # Deadline to fulfill contract
    payment: Payment  # Payment terms
    deliveries: List[Delivery]  # Delivery requirements

    def all_deliveries_fulfilled(self) -> bool:
        """Check if all delivery requirements are met"""
        return all(d.is_fulfilled() for d in self.deliveries)


class Contract:
    """
    Contract entity - represents a SpaceTraders contract

    Invariants:
    - contract_id must be unique and non-empty
    - Cannot accept an already accepted contract
    - Status transitions must be valid
    """

    def __init__(
        self,
        contract_id: str,
        faction_symbol: str,
        type: str,
        terms: ContractTerms,
        accepted: bool,
        fulfilled: bool,
        deadline_to_accept: datetime
    ):
        """
        Initialize a Contract entity

        Args:
            contract_id: Unique contract identifier
            faction_symbol: Faction offering the contract
            type: Contract type (e.g., "PROCUREMENT")
            terms: Contract terms and conditions
            accepted: Whether contract has been accepted
            fulfilled: Whether contract has been fulfilled
            deadline_to_accept: Deadline to accept the contract

        Raises:
            ValueError: If validation fails
        """
        if not contract_id or not contract_id.strip():
            raise ValueError("contract_id cannot be empty")
        if not faction_symbol or not faction_symbol.strip():
            raise ValueError("faction_symbol cannot be empty")

        self._contract_id = contract_id.strip()
        self._faction_symbol = faction_symbol.strip()
        self._type = type
        self._terms = terms
        self._accepted = accepted
        self._fulfilled = fulfilled
        self._deadline_to_accept = deadline_to_accept

    @property
    def contract_id(self) -> str:
        """Get contract ID"""
        return self._contract_id

    @property
    def faction_symbol(self) -> str:
        """Get faction symbol"""
        return self._faction_symbol

    @property
    def type(self) -> str:
        """Get contract type"""
        return self._type

    @property
    def terms(self) -> ContractTerms:
        """Get contract terms"""
        return self._terms

    @property
    def accepted(self) -> bool:
        """Check if contract is accepted"""
        return self._accepted

    @property
    def fulfilled(self) -> bool:
        """Check if contract is fulfilled"""
        return self._fulfilled

    @property
    def deadline_to_accept(self) -> datetime:
        """Get deadline to accept contract"""
        return self._deadline_to_accept

    @property
    def status(self) -> ContractStatus:
        """Get current contract status"""
        if self._fulfilled:
            return ContractStatus.FULFILLED
        elif self._accepted:
            return ContractStatus.ACCEPTED
        else:
            return ContractStatus.OFFERED

    def accept(self) -> None:
        """
        Accept the contract

        Raises:
            ContractAlreadyAcceptedError: If contract is already accepted
        """
        if self._accepted:
            raise ContractAlreadyAcceptedError(
                f"Contract {self._contract_id} is already accepted"
            )
        self._accepted = True

    def is_fulfilled(self) -> bool:
        """
        Check if all contract requirements are fulfilled

        Returns:
            True if all deliveries are complete, False otherwise
        """
        return self._terms.all_deliveries_fulfilled()

    def remaining_units(self) -> int:
        """
        Calculate total remaining units across all deliveries

        Returns:
            Total units still needed to fulfill contract
        """
        return sum(d.remaining() for d in self._terms.deliveries)

    def __repr__(self) -> str:
        return (
            f"Contract(id={self._contract_id}, "
            f"faction={self._faction_symbol}, "
            f"status={self.status.value})"
        )

File: gobot/internal/domain/contract/contract.go
package contract

import (
	"fmt"
	"time"

	"github.com/andrescamacho/spacetraders-go/internal/domain/shared"
)

type Payment struct {
	OnAccepted  int
	OnFulfilled int
}

type Delivery struct {
	TradeSymbol       string
	DestinationSymbol string
	UnitsRequired     int
	UnitsFulfilled    int
}

type Terms struct {
	Payment          Payment
	Deliveries       []Delivery
	DeadlineToAccept string
	Deadline         string
}

type Contract struct {
	contractID    string
	playerID      shared.PlayerID
	factionSymbol string
	contractType  string
	terms         Terms
	accepted      bool
	fulfilled     bool
	clock         shared.Clock
}

// NewContract creates a new contract
// The clock parameter is optional - if nil, defaults to RealClock for production use
func NewContract(contractID string, playerID shared.PlayerID, factionSymbol, contractType string, terms Terms, clock shared.Clock) (*Contract, error) {
	if contractID == "" {
		return nil, fmt.Errorf("contract ID cannot be empty")
	}
	if playerID.IsZero() {
		return nil, fmt.Errorf("invalid player ID")
	}
	if factionSymbol == "" {
		return nil, fmt.Errorf("faction symbol cannot be empty")
	}
	if len(terms.Deliveries) == 0 {
		return nil, fmt.Errorf("contract must have at least one delivery")
	}

	// Default to RealClock if not provided
	if clock == nil {
		clock = shared.NewRealClock()
	}

	return &Contract{
		contractID:    contractID,
		playerID:      playerID,
		factionSymbol: factionSymbol,
		contractType:  contractType,
		terms:         terms,
		accepted:      false,
		fulfilled:     false,
		clock:         clock,
	}, nil
}

func (c *Contract) ContractID() string        { return c.contractID }
func (c *Contract) PlayerID() shared.PlayerID { return c.playerID }
func (c *Contract) FactionSymbol() string     { return c.factionSymbol }
func (c *Contract) Type() string              { return c.contractType }
func (c *Contract) Terms() Terms              { return c.terms }
func (c *Contract) Accepted() bool            { return c.accepted }
func (c *Contract) Fulfilled() bool           { return c.fulfilled }

// Accept accepts the contract (MUTABLE - modifies in place)
func (c *Contract) Accept() error {
	if c.fulfilled {
		return fmt.Errorf("contract already fulfilled")
	}
	if c.accepted {
		return fmt.Errorf("contract already accepted")
	}
	c.accepted = true
	return nil
}

// DeliverCargo updates delivery progress (MUTABLE)
func (c *Contract) DeliverCargo(tradeSymbol string, units int) error {
	if !c.accepted {
		return fmt.Errorf("contract not accepted")
	}

	var delivery *Delivery
	for i := range c.terms.Deliveries {
		if c.terms.Deliveries[i].TradeSymbol == tradeSymbol {
			delivery = &c.terms.Deliveries[i]
			break
		}
	}

	if delivery == nil {
		return fmt.Errorf("trade symbol not in contract")
	}

	if delivery.UnitsFulfilled+units > delivery.UnitsRequired {
		return fmt.Errorf("units exceed required")
	}

	delivery.UnitsFulfilled += units
	return nil
}

// MarkDeliveryTermsMet reconciles a single delivery to fully fulfilled after the
// server has confirmed that good's terms are already met — a SpaceTraders 4509
// "delivery terms ... have been met" response — while the local row had lagged
// behind at a lower UnitsFulfilled. It sets UnitsFulfilled = UnitsRequired for
// ONLY the named good (server truth), so a good the server has NOT confirmed is
// never marked fulfilled and a partial over-fetch attempt does not under-record
// the true server count. Idempotent: re-marking an already-met delivery is a
// no-op. This heals the server-met/local-unmet divergence behind the sp-1pf0r
// contract-fleet wedge.
func (c *Contract) MarkDeliveryTermsMet(tradeSymbol string) error {
	if !c.accepted {
		return fmt.Errorf("contract not accepted")
	}
	for i := range c.terms.Deliveries {
		if c.terms.Deliveries[i].TradeSymbol == tradeSymbol {
			c.terms.Deliveries[i].UnitsFulfilled = c.terms.Deliveries[i].UnitsRequired
			return nil
		}
	}
	return fmt.Errorf("trade symbol not in contract")
}

// ReconcileDeliveredFromServer RAISES each delivery's UnitsFulfilled to the count the GAME
// SERVER reports, and returns whether anything moved. observed is keyed by trade symbol; a
// good the server did not report is left alone.
//
// The local count is a CACHE of a number the server owns: it is written only after a deliver
// the server has already accepted, so a worker that dies between the deliver landing and the
// local write leaves it permanently BEHIND. That is the 2026-08-05 TORWIND double-delivery
// exactly — the local row read 0/47 while the server read 94/47. It delivered 47, crashed
// before recording, resumed from the stale view, and delivered 47 AGAIN.
//
// Raise-only, deliberately, for the same reason ConstructionMaterialTarget.ReconcileDelivered
// is: a LOWER observed value cannot be told apart from a contract read that raced a delivery
// landing between the read and this call, so lowering would erase units that really were
// delivered and re-source them. Raise-only is also the fail-closed direction here (RULINGS #4)
// — it can only ever REDUCE what we go on to buy and deliver, never increase it — and it makes
// the operation monotonic and idempotent under any interleaving with DeliverCargo.
//
// The observed count is stored RAW, uncapped. The server reported 94 against a required 47;
// clamping to 47 would erase the only surviving evidence that the over-delivery happened.
// CanFulfill and every units-remaining computation compare with >= / <=, so an over-count
// reads correctly as "done" everywhere it is consumed.
func (c *Contract) ReconcileDeliveredFromServer(observed map[string]int) bool {
	raised := false
	for i := range c.terms.Deliveries {
		serverUnits, reported := observed[c.terms.Deliveries[i].TradeSymbol]
		if !reported || serverUnits <= c.terms.Deliveries[i].UnitsFulfilled {
			continue
		}
		c.terms.Deliveries[i].UnitsFulfilled = serverUnits
		raised = true
	}
	return raised
}

// MarkAcceptedFromServer records that the server considers this contract accepted, so a local
// row that missed the accept write does not block a fulfil (Fulfill refuses an unaccepted
// contract). Raise-only like the delivery counts above: it never un-accepts.
func (c *Contract) MarkAcceptedFromServer() {
	c.accepted = true
}

// MarkFulfilledFromServer records that the server has ALREADY fulfilled and paid this contract,
// so the local row stops advertising it as active work. Without it, FindActiveContracts
// (accepted AND NOT fulfilled) keeps handing a finished contract to every worker that asks, and
// each one resumes work that can never complete.
//
// Every delivery is raised to its required count first: the server cannot have fulfilled a
// contract whose deliveries it did not consider met, and the persistence layer replays Fulfill()
// when loading the row — which fails its own CanFulfill guard on a row claiming fulfilled with
// short deliveries, making the contract unreadable from that point on.
func (c *Contract) MarkFulfilledFromServer() {
	for i := range c.terms.Deliveries {
		if c.terms.Deliveries[i].UnitsFulfilled < c.terms.Deliveries[i].UnitsRequired {
			c.terms.Deliveries[i].UnitsFulfilled = c.terms.Deliveries[i].UnitsRequired
		}
	}
	c.accepted = true
	c.fulfilled = true
}

// CanFulfill checks if all deliveries are complete
func (c *Contract) CanFulfill() bool {
	for _, delivery := range c.terms.Deliveries {
		if delivery.UnitsFulfilled < delivery.UnitsRequired {
			return false
		}
	}
	return true
}

// Fulfill marks contract as fulfilled (MUTABLE)
func (c *Contract) Fulfill() error {
	if !c.accepted {
		return fmt.Errorf("contract not accepted")
	}
	if !c.CanFulfill() {
		return fmt.Errorf("deliveries not complete")
	}
	c.fulfilled = true
	return nil
}

// IsExpired checks if contract is past deadline
func (c *Contract) IsExpired() bool {
	deadline, err := time.Parse(time.RFC3339, c.terms.Deadline)
	if err != nil {
		return false
	}
	return c.clock.Now().UTC().After(deadline)
}

// EvaluateProfitability delegates profitability calculation to ContractProfitabilityService.
//
// This method provides a convenient API for contract profitability evaluation while
// maintaining proper separation of concerns. The actual business logic resides in
// ContractProfitabilityService.
//
// ctx carries market prices, cargo capacity and fuel costs; it is not a
// context.Context. Errors when a required good has no market price.
func (c *Contract) EvaluateProfitability(ctx ProfitabilityContext) (*ProfitabilityEvaluation, error) {
	service := NewContractProfitabilityService()
	return service.EvaluateProfitability(c, ctx)
}
